lib/codex/utils/mod_utils.dart

```diff
@@ -0,0 +1,138 @@
+import 'dart:async';
+import 'dart:ui' as ui;
+
+import 'package:cached_network_image/cached_network_image.dart';
+import 'package:flutter/material.dart';
+import 'package:warframestat_client/warframestat_client.dart';
+
+const framPath = 'https://cdn.warframestat.us/genesis/modFrames';
+
+typedef ModBackground = ({
+  ui.Image background,
+  ui.Image backer,
+  ui.Image lowerTab
+});
+
+typedef ModFrame = ({
+  ui.Image top,
+  ui.Image sideLights,
+  ui.Image cornerLights,
+  ui.Image bottom,
+});
+
+typedef ModAssets = ({
+  ui.Image thumbnail,
+  ui.Image? polarity,
+  ui.Image rankSlotActive,
+  ui.Image rankSlotEmpty,
+  ModBackground surface,
+  ModFrame frame,
+});
+
+String formatDescription(
+  String? description,
+  List<LevelStat>? levelStats, {
+  int rank = 0,
+}) {
+  if (description != null && description.isNotEmpty) return description;
+
+  final buffer = StringBuffer();
+  if (levelStats != null) {
+    for (final stat in levelStats[rank].stats) {
+      buffer.write('$stat\n');
+    }
+
+    return buffer.toString();
+  }
+
+  return 'Unable to parse mod description';
+}
+
+Future<ui.Image> fetchImage(String part) async {
+  final completer = Completer<ImageInfo>();
+  final image = CachedNetworkImageProvider(part);
+  final listener = ImageStreamListener((info, _) {
+    if (!completer.isCompleted) completer.complete(info);
+  });
+
+  image.resolve(ImageConfiguration.empty).addListener(listener);
+
+  return (await completer.future).image;
+}
+
+String polarityUrl(String polarity) {
+  const polarityPath = 'https://cdn.warframestat.us/genesis/img/polarities';
```

Comment:
oh dart doesn't have objects like that it would be something like

```dart
const imgHost = 'https://cdn.warframestat.us/genesis'

// This is basically a Map<String, dynamic>
const paths = <String, dynamic>{
  polarities: '/img/polarities',
  frames: '/modframes',
};

String polarityUrl(String polarity) {
    return Uri.parse('$imgHost${paths['polarities']}/${polarity.toLowerCase()}.png').toString();
}
```